elements/lisk-p2p/src/peer/outbound.ts

```diff
@@ -0,0 +1,340 @@
+	P2PMessagePacket,
+	P2PNodeInfo,
+	P2PPeerInfo,
+	P2PRequestPacket,
+	P2PResponsePacket,
+} from '../p2p_types';
+
+import * as querystring from 'querystring';
+import * as socketClusterClient from 'socketcluster-client';
+import { validatePeerInfo } from '../validation';
+
+type SCClientSocket = socketClusterClient.SCClientSocket;
+
+export const EVENT_CONNECT_OUTBOUND = 'connectOutbound';
+export const EVENT_CONNECT_ABORT_OUTBOUND = 'connectAbortOutbound';
+export const EVENT_CLOSE_OUTBOUND = 'closeOutbound';
+export const EVENT_OUTBOUND_SOCKET_ERROR = 'outboundSocketError';
+
+export interface PeerInfoAndOutboundConnection {
+	readonly peerInfo: P2PDiscoveredPeerInfo;
+	readonly socket: SCClientSocket;
+}
+
+export class OutboundPeer extends Peer {
+	protected _socket: SCClientSocket | undefined;
+
+	public constructor(
+		peerInfo: P2PDiscoveredPeerInfo,
+		peerSocket?: SCClientSocket,
+		peerConfig?: PeerConfig,
+	) {
+		super(peerInfo, peerConfig);
+		this._socket = peerSocket ? peerSocket : undefined;
+		if (this._socket) {
+			this._bindHandlersToOutboundSocket(this._socket);
+		}
+	}
+
+	public set socket(scClientSocket: SCClientSocket) {
+		if (this._socket) {
+			this._unbindHandlersFromOutboundSocket(this._socket);
+		}
+		this._socket = scClientSocket;
+		this._bindHandlersToOutboundSocket(this._socket);
+	}
+
+	public send(packet: P2PMessagePacket): void {
+		if (!this._socket) {
+			this._socket = this._createOutboundSocket();
+		}
+
+		super.send(packet);
+	}
+
+	public async request(packet: P2PRequestPacket): Promise<P2PResponsePacket> {
+		if (!this._socket) {
+			this._socket = this._createOutboundSocket();
+		}
+
+		return super.request(packet);
+	}
+
+	private _createOutboundSocket(): SCClientSocket {
+		const legacyNodeInfo = this._nodeInfo
+			? convertNodeInfoToLegacyFormat(this._nodeInfo)
+			: undefined;
+
+		const connectTimeout = this._peerConfig.connectTimeout
+			? this._peerConfig.connectTimeout
+			: DEFAULT_CONNECT_TIMEOUT;
+		const ackTimeout = this._peerConfig.ackTimeout
+			? this._peerConfig.ackTimeout
+			: DEFAULT_ACK_TIMEOUT;
+
+		// Ideally, we should JSON-serialize the whole NodeInfo object but this cannot be done for compatibility reasons, so instead we put it inside an options property.
+		const clientOptions: ClientOptionsUpdated = {
+			hostname: this._ipAddress,
+			port: this._wsPort,
+			query: querystring.stringify({
+				...legacyNodeInfo,
+				options: JSON.stringify(legacyNodeInfo),
+			}),
+			connectTimeout,
+			ackTimeout,
+			multiplex: false,
+			autoConnect: false,
+			autoReconnect: false,
+			pingTimeoutDisabled: true,
+		};
+
+		const outboundSocket = socketClusterClient.create(clientOptions);
+
+		this._bindHandlersToOutboundSocket(outboundSocket);
+
+		return outboundSocket;
+	}
+
+	public connect(): void {
+		if (!this._socket) {
+			this._socket = this._createOutboundSocket();
+		}
+		this._socket.connect();
+	}
+
+	public disconnect(code: number = 1000, reason?: string): void {
```

Comment:
it seems like there are lots of redundant code between outbound/inbound peer, why not put it in base?

Reply:
There is not so much. We rewrite the base methods that are different depending on the kind of peer. We decided to create two different classes to have a better view of every kind. 

Besides tslint doesn't allow more than one class per file.